src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/Cohost/CohostUriPresentationEndpoint.cs

```diff
@@ -0,0 +1,103 @@
+
+#pragma warning disable RS0030 // Do not use banned APIs
+[Shared]
+[CohostEndpoint(VSInternalMethods.TextDocumentUriPresentationName)]
+[Export(typeof(IDynamicRegistrationProvider))]
+[ExportCohostStatelessLspService(typeof(CohostUriPresentationEndpoint))]
+[method: ImportingConstructor]
+#pragma warning restore RS0030 // Do not use banned APIs
+internal class CohostUriPresentationEndpoint(IRemoteClientProvider remoteClientProvider, ILoggerFactory loggerFactory)
+    : AbstractRazorCohostDocumentRequestHandler<UriPresentationParams, WorkspaceEdit?>, IDynamicRegistrationProvider
+{
+    private readonly IRemoteClientProvider _remoteClientProvider = remoteClientProvider;
+    private readonly ILogger _logger = loggerFactory.GetOrCreateLogger<CohostUriPresentationEndpoint>();
+
+    protected override bool MutatesSolutionState => false;
+
+    protected override bool RequiresLSPSolution => true;
+
+    public Registration? GetRegistration(VSInternalClientCapabilities clientCapabilities, DocumentFilter[] filter, RazorCohostRequestContext requestContext)
+    {
+        if (clientCapabilities.SupportsVisualStudioExtensions)
+        {
+            return new Registration
+            {
+                Method = VSInternalMethods.TextDocumentUriPresentationName,
+                RegisterOptions = new TextDocumentRegistrationOptions()
+                {
+                    DocumentSelector = filter
+                }
+            };
+        }
+
+        return null;
+    }
+
+    protected override RazorTextDocumentIdentifier? GetRazorTextDocumentIdentifier(UriPresentationParams request)
+        => request.TextDocument.ToRazorTextDocumentIdentifier();
+
+    protected override async Task<WorkspaceEdit?> HandleRequestAsync(UriPresentationParams request, RazorCohostRequestContext context, CancellationToken cancellationToken)
+    {
+        var razorDocument = context.TextDocument.AssumeNotNull();
+
+        var remoteClient = await _remoteClientProvider.TryGetClientAsync(cancellationToken).ConfigureAwait(false);
+        if (remoteClient is null)
+        {
+            _logger.LogWarning($"Couldn't get remote client");
+            return null;
+        }
+
+        try
+        {
+            var data = await remoteClient.TryInvokeAsync<IRemoteUriPresentationService, TextChange?>(
+                razorDocument.Project.Solution,
+                (service, solutionInfo, cancellationToken) => service.GetPresentationAsync(solutionInfo, razorDocument.Id, request.Range.ToLinePositionSpan(), request.Uris, cancellationToken),
+                cancellationToken).ConfigureAwait(false);
+
+            if (data.Value is { } textChange)
+            {
+                var sourceText = await razorDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+                return new WorkspaceEdit
+                {
+                    DocumentChanges = new TextDocumentEdit[]
+                    {
```

Comment:
Nit: we could use new fancy array initializer syntax here []

Reply:
which you use below :)

Reply:
Ah, OK.

Reply:
In theory they are but only if the compiler can determine what the type of the array should be. If this is a SumType, then it can't so we have to tell it.